File: java/L100/LeetCode0131.java
package L100;

import java.util.ArrayList;
import java.util.List;

public class LeetCode0131 {
    private List<List<String>> res;
    private int N;
    public List<List<String>> partition(String s) {
        res = new ArrayList<>();
        N = s.length();
        List<String> path = new ArrayList<>(N);
        trace(s, 0, path);
        return res;
    }

    private void trace(String s, int index, List<String> path) {
        if (index >= N) {
            res.add(new ArrayList<>(path));
            return;
        }
        for (int i = index;i < N;++i) {
            if (isPalindrome(s, index, i)) {
                path.add(s.substring(index, i + 1));
                trace(s, i + 1, path);
                path.remove(path.size() - 1);
            }
        }
    }

    private boolean isPalindrome(String s, int l, int r) {
        while (l < r) {
            if (s.charAt(l) != s.charAt(r)) {
                return false;
            }
            ++l;
            --r;
        }
        return true;
    }
}

File: cpp/L100/LeetCode0131.cc
#include <string>
using std::string;
#include <vector>
using std::vector;

class Solution {
    bool isPalindrome(string const &s, int left, int right) {
        for (;left < right;++left, --right) {
            if (s[left] != s[right]) {
                return false;
            }
        }
        return true;
    }
    void dfs(string const &s, int index, vector<vector<string>> &result, vector<string> &path) {
        if (index >= s.size()) {
            result.emplace_back(vector<string>(path.begin(), path.end()));
        } else {
            for (int i{index};i < s.size();++i) {
                if (isPalindrome(s, index, i)) {
                    path.emplace_back(s.substr(index, i - index + 1));
                    dfs(s, i + 1, result, path);
                    path.pop_back();
                }
            }
        }
    }
public:
    vector<vector<string>> partition(string s) {
        vector<vector<string>> result;
        vector<string> path;
        dfs(s, 0, result, path);
        return result;
    }
};